--- a/qmail-smtpd.c	2004-05-02 19:14:41.000000000 +0300
+++ b/qmail-smtpd.c	2004-05-02 19:14:14.000000000 +0300
@@ -120,6 +120,7 @@
 void err_nogwcert() { out("553 no valid cert for gatewaying (#5.7.1)\r\n"); }
 #endif
 void err_unimpl() { out("502 unimplemented (#5.5.1)\r\n"); }
+void err_norouting() { out("553 mail routing not allowed\r\n"); }
 void err_syntax() { out("555 syntax error (#5.5.4)\r\n"); }
 void err_wantmail() { out("503 MAIL first (#5.5.1)\r\n"); }
 void err_wantrcpt() { out("503 RCPT first (#5.5.1)\r\n"); }
@@ -351,6 +352,10 @@
   return r;
 }
 
+int wantsrouting()
+{
+  return (strchr (addr.s, '%') || strchr (addr.s, '!'));
+}
 
 int seenmail = 0;
 int flagbarfbmf; /* defined if seenmail */
@@ -417,6 +422,7 @@
     err_bmt();
     return;
   }
+  if (wantsrouting()) { err_norouting(); return; }
   if (relayclient) {
     --addr.len;
     if (!stralloc_cats(&addr,relayclient)) die_nomem();
