diff -urN qmail-1.03/rblsmtpd-0.70/rblsmtpd.c qmail-1.03.new/rblsmtpd-0.70/rblsmtpd.c
--- qmail-1.03/rblsmtpd-0.70/rblsmtpd.c	2004-04-25 14:06:16.013461600 +0200
+++ qmail-1.03.new/rblsmtpd-0.70/rblsmtpd.c	2004-04-25 14:10:54.203170336 +0200
@@ -32,19 +32,19 @@
 char ssoutbuf[1];
 substdio ssout = SUBSTDIO_FDBUF(write,1,ssoutbuf,sizeof ssoutbuf);
 
-void reject() { substdio_putflush(&ssout,message.s,message.len); }
-void accept() { substdio_putsflush(&ssout,"250 rblsmtpd.local\r\n"); }
-void greet()  { substdio_putsflush(&ssout,"220 rblsmtpd.local\r\n"); }
-void quit()   { substdio_putsflush(&ssout,"221 rblsmtpd.local\r\n"); _exit(0); }
+void reject_msg() { substdio_putflush(&ssout,message.s,message.len); }
+void accept_msg() { substdio_putsflush(&ssout,"250 rblsmtpd.local\r\n"); }
+void greet_msg()  { substdio_putsflush(&ssout,"220 rblsmtpd.local\r\n"); }
+void quit_msg()   { substdio_putsflush(&ssout,"221 rblsmtpd.local\r\n"); _exit(0); }
 
 struct commands smtpcommands[] = {
-  { "quit", quit, 0 }
-, { "helo", accept, 0 }
-, { "ehlo", accept, 0 }
-, { "mail", accept, 0 }
-, { "rset", accept, 0 }
-, { "noop", accept, 0 }
-, { 0, reject, 0 }
+  { "quit", quit_msg, 0 }
+, { "helo", accept_msg, 0 }
+, { "ehlo", accept_msg, 0 }
+, { "mail", accept_msg, 0 }
+, { "rset", accept_msg, 0 }
+, { "noop", accept_msg, 0 }
+, { 0, reject_msg, 0 }
 } ;
 
 void check()
@@ -99,10 +99,10 @@
      }
   if (!stralloc_cats(&message,"\r\n")) die_sys();
 
-  if (!timeout) { reject(); _exit(0); }
+  if (!timeout) { reject_msg(); _exit(0); }
   alarm(timeout);
 
-  greet();
+  greet_msg();
   commands(&ssin,smtpcommands);
   die_sys();
 }
