diff -ur rblsmtpd-0.50.orig/conf-home rblsmtpd-0.50/conf-home
--- rblsmtpd-0.70.orig/conf-home	Thu Aug 13 11:35:36 1998
+++ rblsmtpd-0.70/conf-home	Fri Aug 14 13:14:15 1998
@@ -1,4 +1,4 @@
-/usr/local
+/var/qmail
 
 This is the rblsmtpd home directory. rblsmtpd will be installed in
 .../bin.
Only in rblsmtpd-0.70: new.patch
diff -ur rblsmtpd-0.70.orig/rblsmtpd.8 rblsmtpd-0.70/rblsmtpd.8
--- rblsmtpd-0.70.orig/rblsmtpd.8	Tue Aug 25 10:58:22 1998
+++ rblsmtpd-0.70/rblsmtpd.8	Wed Aug 26 20:10:59 1998
@@ -6,6 +6,8 @@
 [
 .B \-b
 ] [
+.B \-l
+] [
 .B \-R
 ] [
 .B \-r \fIdomain
@@ -32,7 +34,7 @@
 if it is told to block mail from this client.
 Instead it carries out its own limited SMTP conversation,
 temporarily rejecting all attempts to send a message.
-Meanwhile it prints one line on
+Meanwhile, if the -l flag is used, it prints one line on
 .B stderr
 to log its activity.
 
diff -ur rblsmtpd-0.70.orig/rblsmtpd.c rblsmtpd-0.70/rblsmtpd.c
--- rblsmtpd-0.70.orig/rblsmtpd.c	Tue Aug 25 10:58:22 1998
+++ rblsmtpd-0.70/rblsmtpd.c	Wed Aug 26 19:59:26 1998
@@ -15,6 +15,7 @@
 
 int flagbounce = 0;
 int flagrblsafe = 0;
+int log = 0;
 char *rbldomain = "rbl.maps.vix.com";
 unsigned int timeout = 60;
 
@@ -83,12 +84,15 @@
   for (i = 0;i < message.len;++i)
     if ((message.s[i] < 32) || (message.s[i] > 126))
       message.s[i] = '?';
+  if ( log )
+     {
   substdio_puts(subfderr,"rblsmtpd: pid ");
   substdio_put(subfderr,strnum,fmt_ulong(strnum,(unsigned long) getpid()));
   substdio_puts(subfderr,": ");
   substdio_put(subfderr,message.s,message.len);
   substdio_puts(subfderr,"\n");
   substdio_flush(subfderr);
+     }
   if (!stralloc_cats(&message,"\r\n")) die_sys();
 
   if (!timeout) { reject(); _exit(0); }
@@ -101,7 +105,7 @@
 
 void usage()
 {
-  strerr_die1x(100,"rblsmtpd: usage: rblsmtpd [ -b ] [ -R ] [ -r domain ] [ -t timeout ] smtpd [ arg ... ]");
+  strerr_die1x(100,"rblsmtpd: usage: rblsmtpd [ -b ] [ -l ] [ -R ] [ -r domain ] [ -t timeout ] smtpd [ arg ... ]");
 }
 
 void main(argc,argv)
@@ -110,12 +114,13 @@
 {
   int opt;
 
-  while ((opt = getopt(argc,argv,"t:r:Rb")) != opteof)
+  while ((opt = getopt(argc,argv,"t:r:Rbl")) != opteof)
     switch(opt) {
       case 't': scan_uint(optarg,&timeout); break;
       case 'R': flagrblsafe = 1; break;
       case 'b': flagbounce = 1; break;
       case 'r': rbldomain = optarg; break;
+      case 'l': log = 1; break;
       default: usage();
     }
   argv += optind;
