diff -urN qmail-1.03.orig/rblsmtpd-0.70/Makefile qmail-1.03/rblsmtpd-0.70/Makefile
--- qmail-1.03.orig/rblsmtpd-0.70/Makefile	Mon Mar  6 05:10:23 2000
+++ qmail-1.03/rblsmtpd-0.70/Makefile	Mon Mar  6 05:32:04 2000
@@ -128,10 +128,18 @@
 compile fmt_ulong.c fmt.h
 	./compile fmt_ulong.c
 
+fmt_uint.o: \
+compile fmt_uint.c fmt.h
+	./compile fmt_uint.c
+
+fmt_uint0.o: \
+compile fmt_uint0.c fmt.h
+	./compile fmt_uint0.c
+
 fs.a: \
-makelib fmt_str.o fmt_ulong.o scan_uint.o scan_ulong.o scan_xlong.o
-	./makelib fs.a fmt_str.o fmt_ulong.o scan_uint.o \
-	scan_ulong.o scan_xlong.o
+makelib fmt_str.o fmt_ulong.o scan_uint.o scan_ulong.o fmt_uint.o fmt_uint0.o scan_xlong.o
+	./makelib fs.a fmt_str.o fmt_ulong.o fmt_uint.o fmt_uint0.o \
+	scan_uint.o scan_ulong.o scan_xlong.o
 
 getopt.a: \
 makelib subgetopt.o sgetopt.o
@@ -213,9 +221,9 @@
 rblsmtpd antirbl
 
 rblsmtpd: \
-load rblsmtpd.o txt.o commands.o ip.o getopt.a strerr.a substdio.a \
+load rblsmtpd.o txt.o commands.o ip.o logger.o getopt.a strerr.a substdio.a \
 stralloc.a alloc.a error.a case.a env.a str.a fs.a dns.lib socket.lib
-	./load rblsmtpd txt.o commands.o ip.o getopt.a strerr.a \
+	./load rblsmtpd txt.o commands.o ip.o logger.o getopt.a strerr.a \
 	substdio.a stralloc.a alloc.a error.a case.a env.a str.a \
 	fs.a  `cat dns.lib` `cat socket.lib`
 
@@ -370,3 +378,7 @@
 txt.o: \
 compile txt.c stralloc.h gen_alloc.h str.h fmt.h ip.h
 	./compile txt.c
+
+logger.o: \
+compile logger.c error.h substdio.h subfd.h exit.h str.h scan.h fmt.h
+	./compile logger.c
diff -urN qmail-1.03.orig/rblsmtpd-0.70/fmt_uint.c qmail-1.03/rblsmtpd-0.70/fmt_uint.c
--- qmail-1.03.orig/rblsmtpd-0.70/fmt_uint.c	Thu Jan  1 01:00:00 1970
+++ qmail-1.03/rblsmtpd-0.70/fmt_uint.c	Mon Jun 15 12:53:16 1998
@@ -0,0 +1,6 @@
+#include "fmt.h"
+
+unsigned int fmt_uint(s,u) register char *s; register unsigned int u;
+{
+  register unsigned long l; l = u; return fmt_ulong(s,l);
+}
Binary files qmail-1.03.orig/rblsmtpd-0.70/fmt_uint.o and qmail-1.03/rblsmtpd-0.70/fmt_uint.o differ
diff -urN qmail-1.03.orig/rblsmtpd-0.70/fmt_uint0.c qmail-1.03/rblsmtpd-0.70/fmt_uint0.c
--- qmail-1.03.orig/rblsmtpd-0.70/fmt_uint0.c	Thu Jan  1 01:00:00 1970
+++ qmail-1.03/rblsmtpd-0.70/fmt_uint0.c	Mon Jun 15 12:53:16 1998
@@ -0,0 +1,10 @@
+#include "fmt.h"
+
+unsigned int fmt_uint0(s,u,n) char *s; unsigned int u; unsigned int n;
+{
+  unsigned int len;
+  len = fmt_uint(FMT_LEN,u);
+  while (len < n) { if (s) *s++ = '0'; ++len; }
+  if (s) fmt_uint(s,u);
+  return len;
+}
diff -urN qmail-1.03.orig/rblsmtpd-0.70/logger.c qmail-1.03/rblsmtpd-0.70/logger.c
--- qmail-1.03.orig/rblsmtpd-0.70/logger.c	Thu Jan  1 01:00:00 1970
+++ qmail-1.03/rblsmtpd-0.70/logger.c	Mon Mar  6 05:46:25 2000
@@ -0,0 +1,58 @@
+#include <sys/types.h>
+#include <sys/time.h>
+#include <syslog.h>
+#include "error.h"
+#include "substdio.h"
+#include "subfd.h"
+#include "exit.h"
+#include "str.h"
+#include "scan.h"
+#include "fmt.h"
+
+char buf[800]; /* syslog truncates long lines (or crashes); GPACIC */
+int bufpos = 0; /* 0 <= bufpos < sizeof(buf) */
+int flagcont = 0;
+int priority; /* defined if flagcont */
+char stamp[FMT_ULONG + FMT_ULONG + 3]; /* defined if flagcont */
+
+void stamp_make()
+{
+  struct timeval tv;
+  char *s;
+  gettimeofday(&tv,(struct timezone *) 0);
+  s = stamp;
+  s += fmt_ulong(s,(unsigned long) tv.tv_sec);
+  *s++ = '.';
+  s += fmt_uint0(s,(unsigned int) tv.tv_usec,6);
+  *s = 0;
+}
+
+void flush()
+{
+  if (bufpos) {
+    buf[bufpos] = 0;
+    if (flagcont)
+      syslog(priority,"%s+%s",stamp,buf); /* logger folds invisibly; GPACIC */
+    else {
+      stamp_make();
+      priority = LOG_INFO;
+      if (str_start(buf,"warning:")) priority = LOG_WARNING;
+      if (str_start(buf,"alert:")) priority = LOG_ALERT;
+      syslog(priority,"%s %s",stamp,buf);
+      flagcont = 1;
+    }
+  }
+  bufpos = 0;
+}
+
+void rbl_syslog(s,len)
+char *s;
+int len;
+{
+  while (len--) {
+    if (*s == '\n') { flush(); flagcont = 0; s++; continue; }
+    if (bufpos == sizeof(buf) - 1) flush();
+    if ((*s < 32) || (*s > 126)) *s = '?'; /* logger truncates at 0; GPACIC */
+    buf[bufpos++] = *s++;
+  }
+}
diff -urN qmail-1.03.orig/rblsmtpd-0.70/rblsmtpd.c qmail-1.03/rblsmtpd-0.70/rblsmtpd.c
--- qmail-1.03.orig/rblsmtpd-0.70/rblsmtpd.c	Mon Mar  6 05:10:23 2000
+++ qmail-1.03/rblsmtpd-0.70/rblsmtpd.c	Mon Mar  6 05:47:08 2000
@@ -1,3 +1,4 @@
+#include <syslog.h>
 #include "substdio.h"
 #include "subfd.h"
 #include "readwrite.h"
@@ -92,12 +93,8 @@
       message.s[i] = '?';
   if ( log )
      {
-  substdio_puts(subfderr,"rblsmtpd: pid ");
-  substdio_put(subfderr,strnum,fmt_ulong(strnum,(unsigned long) getpid()));
-  substdio_puts(subfderr,": ");
-  substdio_put(subfderr,message.s,message.len);
-  substdio_puts(subfderr,"\n");
-  substdio_flush(subfderr);
+  rbl_syslog(message.s,message.len);
+  rbl_syslog("\n",1);
      }
   if (!stralloc_cats(&message,"\r\n")) die_sys();
 
@@ -131,6 +128,8 @@
     }
   argv += optind;
   if (!*argv) usage();
+
+  openlog("rblsmtpd",0,LOG_MAIL);
 
   res_init();
   check();
