diff -ur rblsmtpd-0.70.orig/Makefile rblsmtpd-0.70/Makefile
--- rblsmtpd-0.70.orig/Makefile	Tue Aug 25 10:58:22 1998
+++ rblsmtpd-0.70/Makefile	Tue Sep  1 15:24:16 1998
@@ -217,7 +217,7 @@
 stralloc.a alloc.a error.a case.a env.a str.a fs.a dns.lib socket.lib
 	./load rblsmtpd txt.o commands.o ip.o getopt.a strerr.a \
 	substdio.a stralloc.a alloc.a error.a case.a env.a str.a \
-	fs.a  `cat dns.lib` `cat socket.lib`
+	fs.a  `cat dns.lib` `cat socket.lib` -lresolv
 
 rblsmtpd.0: \
 rblsmtpd.8
diff -ur rblsmtpd-0.70.orig/rblsmtpd.c rblsmtpd-0.70/rblsmtpd.c
--- rblsmtpd-0.70.orig/rblsmtpd.c	Tue Aug 25 10:58:22 1998
+++ rblsmtpd-0.70/rblsmtpd.c	Tue Sep  1 15:22:52 1998
@@ -1,3 +1,6 @@
+typedef unsigned short sa_family_t;
+#include <linux/in.h>
+#include <resolv.h>
 #include "substdio.h"
 #include "subfd.h"
 #include "readwrite.h"
diff -ur rblsmtpd-0.70.orig/txt.c rblsmtpd-0.70/txt.c
--- rblsmtpd-0.70.orig/txt.c	Tue Aug 25 10:58:22 1998
+++ rblsmtpd-0.70/txt.c	Tue Sep  1 15:22:18 1998
@@ -1,7 +1,8 @@
 #include <stdio.h>
 #include <netdb.h>
 #include <sys/types.h>
-#include <netinet/in.h>
+typedef unsigned short sa_family_t;
+#include <linux/in.h>
 #include <arpa/nameser.h>
 #include <resolv.h>
 #include <errno.h>
