--- qmail-1.03/rblsmtpd-0.70/tryrsolv.c~	Sat Mar 24 21:33:16 2001
+++ qmail-1.03/rblsmtpd-0.70/tryrsolv.c	Sat Mar 24 21:33:41 2001
@@ -1,7 +1,7 @@
 main()
 {
   dn_expand();
-  res_init();
+  __res_init();
   res_query();
   res_search();
 }
--- qmail-1.03/rblsmtpd-0.70/rblsmtpd.c~	Sat Mar 24 21:35:26 2001
+++ qmail-1.03/rblsmtpd-0.70/rblsmtpd.c	Sat Mar 24 21:36:27 2001
@@ -1,4 +1,5 @@
 #include <syslog.h>
+#include <resolv.h>
 #include "substdio.h"
 #include "subfd.h"
 #include "readwrite.h"
