--- qmail-smtpd.c.orig	Mon Jun 15 06:53:16 1998
+++ qmail-smtpd.c	Sat Feb 13 11:35:23 1999
@@ -27,6 +27,8 @@
 #define MAXHOPS 100
 unsigned int databytes = 0;
 int timeout = 1200;
+int tarpitcount = 0;
+int tarpitdelay = 5;
 
 int safewrite(fd,buf,len) int fd; char *buf; int len;
 {
@@ -110,6 +112,13 @@
   if (control_readint(&timeout,"control/timeoutsmtpd") == -1) die_control();
   if (timeout <= 0) timeout = 1;
 
+  if (!env_get("NOTARPIT")) {
+    if (control_readint(&tarpitcount,"control/tarpitcount") == -1) die_control();
+    if (tarpitcount < 0) tarpitcount = 0;
+    if (control_readint(&tarpitdelay,"control/tarpitdelay") == -1) die_control();
+    if (tarpitdelay < 0) tarpitdelay = 0;
+  }
+
   if (rcpthosts_init() == -1) die_control();
 
   bmfok = control_readfile(&bmf,"control/badmailfrom",0);
@@ -221,6 +230,7 @@
 int flagbarf; /* defined if seenmail */
 stralloc mailfrom = {0};
 stralloc rcptto = {0};
+int rcptcount;
 
 void smtp_helo(arg) char *arg;
 {
@@ -245,6 +255,7 @@
   if (!stralloc_copys(&rcptto,"")) die_nomem();
   if (!stralloc_copys(&mailfrom,addr.s)) die_nomem();
   if (!stralloc_0(&mailfrom)) die_nomem();
+  rcptcount = 0;
   out("250 ok\r\n");
 }
 void smtp_rcpt(arg) char *arg; {
@@ -261,6 +272,7 @@
   if (!stralloc_cats(&rcptto,"T")) die_nomem();
   if (!stralloc_cats(&rcptto,addr.s)) die_nomem();
   if (!stralloc_0(&rcptto)) die_nomem();
+  if (tarpitcount && ++rcptcount >= tarpitcount) sleep(tarpitdelay); 
   out("250 ok\r\n");
 }
 
